app/models/fact_value.rb

```diff
@@ -81,13 +81,16 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
+  def self.build_facts_hash(facts, include_host = true)
     hash = {}
     facts.each do |fact|
-      hash[hosts.detect {|h| h.id == fact.host_id}.to_s] ||= {}
-      hash[hosts.detect {|h| h.id == fact.host_id}.to_s].update({fact.name.to_s => fact.value})
+      if include_host
+        hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
```

Comment:
This is making a `.where` call for every fact, we don't need this either, just once at the beginning is fine. Remember it's only needed when `include_host`